tests/integration/test_dnf.py

```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Integration tests for dnf charm library."""
+
+import pathlib
+import unittest
+
+from cleantest.control import Env
+from cleantest.data import File
+from cleantest.provider import LXDArchon, lxd
```

Comment:
I'm a little nervous of introducing this **right now**. Clever though it is, cleantest is relatively young, and also not used anywhere else in this project.

Given that all of the useful tests in this file use AlmaLinux 9, perhaps we could just use the AlmaLinux [cloud images for LXD](https://uk.lxd.images.canonical.com/), and take a similar approach to the existing functional tests that run on Ubuntu?

Reply:
Just to clarify, by "useful tests" I mean that we probably don't need to bother firing up a Ubuntu LXD instance just to validate that `dnf` doesn't work - I'd save yourself the effort/code!

Reply:
I think at this point it's just a little **too new** and unsupported for me to accept it. Remember that this repo supports libraries used by the whole charm ecosystem, and we need to ensure that the tests can be run reliably, without relying on your goodwill to maintain `cleantest` :)

I think for now, we should mimic the behaviour in the original tests, please. Happy for you to open an issue and we can investigate how to simplify the tox file (and multiple environments) in a separate thread - and perhaps one day that *could* involve cleantest - but only after a more thorough evaluation.

@benhoyt any parting words here on this topic?

Reply:
That's a fair assessment. I already opened this issue [here](https://github.com/canonical/operator-libs-linux/issues/76) that pertains to improving the tests on this repo. It circles back to a discussion that Ben and I had somewhere in this thread :sweat_smile:

I will make the dnf tests inline with the current integration tests. Perhaps we can have a more in-depth discussion about cleantest at the next sprint or sometime sooner :smile:

Reply:
Yeah, I think it's wise to at least do a separate assessment. Though I don't mind using a simple library/repo like this as a bit of a test bed ... but let's at least leave to a future discussion/PR.